File: src/EmployeeTasks/Pila.java
package EmployeeTasks;
import java.util.Iterator;
import java.util.NoSuchElementException;

public class Pila<T extends InterfazID> implements Iterable<T> {
    private Nodo<T> tope;
    private int tamanio;
    private int capacidad;

    public int getCapacidad() {
        return capacidad;
    }

    public Pila() {
        this.tope = null;
        this.tamanio = 0;
    }

    public void push(T elemento) {
        Nodo<T> nuevoNodo = new Nodo<>(elemento);
        nuevoNodo.setSiguiente(tope);
        tope = nuevoNodo;
        tamanio++;
    }

    public T pop() {
        if (tope == null) {
            System.out.println("La pila está vacía.");
            return null;
        }
        T eliminado = tope.getElemento();
        tope = tope.getSiguiente();
        tamanio--;
        return eliminado;
    }

    public T peek() {
        if (tope == null) {
            System.out.println("La pila está vacía.");
            return null;
        }
        return tope.getElemento();
    }

    public T buscarPorId(String id) {
        Nodo<T> actual = tope;

        while (actual != null) {
            if (actual.getElemento().getId().equals(id)) {
                return actual.getElemento();
            }
            actual = actual.getSiguiente();
        }
        return null;
    }

    public boolean eliminarPorId(String id) {
        if (tope == null) {
            System.out.println("La pila está vacía.");
            return false;
        }

        Nodo<T> actual = tope;
        Nodo<T> previo = null;

        while (actual != null) {
            if (actual.getElemento().getId().equals(id)) {
                if (previo == null) {
                    tope = actual.getSiguiente();
                }

                else {
                    previo.setSiguiente(actual.getSiguiente());
                }
                tamanio--;
                return true;
            }
            previo = actual;
            actual = actual.getSiguiente();
        }
        return false;
    }

    public int getTamanio() {
        return tamanio;
    }
    
    public void mostrarElementos() {
    if (tope == null) {
        System.out.println("La pila está vacía.");
        return;
    }
    
    System.out.println("Elementos en la pila:");
    Nodo<T> actual = tope;
    
    while (actual != null) {
        System.out.println("- " + actual.getElemento().getId());
        actual = actual.getSiguiente();
    }
}

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private Nodo<T> actual = tope;

            @Override
            public boolean hasNext() {
                return actual != null;
            }

            @Override
            public T next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                T elemento = actual.getElemento();
                actual = actual.getSiguiente();
                return elemento;
            }
        };
    }
}

File: Códigos/TareasEmpleado/Pila.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProyectoEstructuras.Códigos.TareasEmpleado
{
    public class Pila<T> where T : InterfazID
    {
        private Nodo<T> tope;
        private int tamanio;
        private int capacidad;

        public Nodo<T> Tope { get => tope; set => tope = value; }
        public int Tamanio { get => tamanio; set => tamanio = value; }
        public int Capacidad { get => capacidad; set => capacidad = value; }

        public Pila()
        {
            this.tope = null;
            this.tamanio = 0;
        }

        public void Push(T elemento)
        {
            Nodo<T> nuevoNodo = new Nodo<T>(elemento);
            nuevoNodo.Siguiente = tope;
            tope = nuevoNodo;
            tamanio++;
        }

        public T Pop()
        {
            if (tope == null)
            {
                Console.WriteLine("La pila está vacía.");
                return default;
            }

            T eliminado = tope.Elemento;
            tope = tope.Siguiente;
            tamanio--;
            return eliminado;
        }

        public T Peek()
        {
            if (tope == null)
            {
                Console.WriteLine("La pila está vacía.");
                return default;
            }

            return tope.Elemento;
        }

        public T BuscarPorId(string id)
        {
            Nodo<T> actual = tope;

            while (actual != null)
            {
                if (actual.Elemento.GetId().Equals(id))
                {
                    return actual.Elemento;
                }

                actual = actual.Siguiente;
            }

            return default;
        }

        public bool EliminarPorId(string id)
        {
            if (tope == null)
            {
                Console.WriteLine("La pila está vacía.");
                return false;
            }

            Nodo<T> actual = tope;
            Nodo<T> previo = null;

            while (actual != null)
            {
                if (actual.Elemento.GetId().Equals(id))
                {
                    if (previo == null)
                    {
                        tope = actual.Siguiente;
                    }
                    else
                    {
                        previo.Siguiente = actual.Siguiente;
                    }

                    tamanio--;
                    return true;
                }

                previo = actual;
                actual = actual.Siguiente;
            }

            return false;
        }

        public void MostrarElementos()
        {
            if (tope == null)
            {
                Console.WriteLine("La pila está vacía.");
                return;
            }

            Console.WriteLine("Elementos en la pila:");
            Nodo<T> actual = tope;

            while (actual != null)
            {
                Console.WriteLine("- " + actual.Elemento.GetId());
                actual = actual.Siguiente;
            }
        }

    }
}
